cc/testing/evaluation/algorithms_with_insufficient_noise.cc

```diff
@@ -0,0 +1,187 @@
+
+    bool algo_is_dp = tester.Run(); // false = 0, true = 1
+    std::cout << algo_is_dp << std::endl; 
+    return algo_is_dp; // correct outcome is false, e.g., DP premise rejected
+  }
+
+  else {
+    std::cout << "The algorithm specified was not recognized." << 
+      "Try count, boundedsum, or boundedmean instead." << std::endl;
+  }
+}
+
+
+void GetTestResults(std::ofstream& datafile, std::string algorithm,
+  double num_datasets, double num_samples_per_histogram, double ratio_min,
+  double ratio_max) {
+
+  double num_tests = 0;
+  double num_tests_passed = 0;
+  double maximum_ratio_passed = 0;
+  auto start = std::chrono::high_resolution_clock::now();
+
+  for (double i=ratio_min; i<=ratio_max; i++) {
+    auto start_test_run = std::chrono::high_resolution_clock::now();
+    double ratio = i/100.0;
+    std::cout << "Now calculating " << algorithm << " algorithm with ratio: " << 
+      ratio << std::endl;
+    bool outcome = RunStochasticTester(algorithm, ratio, num_datasets,
+      num_samples_per_histogram);
+    num_tests++;
+
+    if (outcome == 0) { // correct outcome is false, e.g., DP predicate rejected 
+      num_tests_passed++;
+      maximum_ratio_passed = ratio;
+    }
+
+    auto finish_test_run = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double> test_run_length =
+      std::chrono::duration_cast<std::chrono::duration<double>>(
+      finish_test_run - start_test_run);
+    datafile << "insufficient_noise," << algorithm << ",0," << outcome << "," <<
+      ratio << "," << num_datasets << "," << num_samples_per_histogram << "," <<
+      test_run_length.count() << "\n";
+  }
+
+  auto end = std::chrono::high_resolution_clock::now();
+  std::chrono::duration<double> time_elapsed =
+    std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
+  std::cout << num_tests_passed << " out of " << num_tests << " passed. " <<
+    "The largest ratio value that passed was " << maximum_ratio_passed << "." <<
+    std::endl;
+    }
+  } // namespace testing
+} // namespace differential_privacy
+
+int main(int argc, char *argv[]) {
+
+  double num_datasets = 15;
+  double num_samples_per_histogram = 1000000;
+  double ratio_min = 80.0;
+  double ratio_max = 99.0;
+  std::ofstream datafile;
+  std::vector<std::string> algorithms={"count","boundedsum","boundedmean"};
+
+    if (argc >= 2) {
+      datafile.open(argv[1]);
+    }
+
+    else {
+      datafile.open("stochastic_tester_results.txt");
+    }
+
+  datafile << "test_name,algorithm,expected,actual,ratio,num_datasets," <<
+    "num_samples,time(sec)" << "\n";
+
+  for (std::string& algorithm : algorithms) {
+    // run the test on each algorithm 10 times
+    for (int i=1; i<=3; i++) {
+      differential_privacy::testing::GetTestResults(datafile,algorithm,
+        num_datasets,num_samples_per_histogram,ratio_min,ratio_max);
+      std::cout << "Number of iterations completed: " << i << "/10" << std::endl;
```

Comment:
The code comment doesn't match the code itself. It looks like only 3 runs per algorithm are happening.

Also, should the number of iterations be a parameter that can be set (like datafile, num_datasets, etc.)?

Reply:
I'm scrapping the idea to run it multiple times, given how long a single run takes.